cloudevents/README.md

```diff
@@ -0,0 +1,145 @@
+# Cloudevents Clients
+
+We have implemented the [cloudevents](https://cloudevents.io/)-based clients in this package to assist developers in
+easily implementing the [Event Based Manifestwork](https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/224-event-based-manifestwork)
+proposal.
+
+## Generic Clients
+
+The generic client (`generic.CloudEventsClient`) is used to resync/publish/subscribe resource objects between sources
+and agents with cloudevents.
+
+A resource object can be any object that implements the `generic.ResourceObject` interface.
+
+### Building a generic client on a source
+
+Developers can use `generic.NewCloudEventSourceClient` method to build a generic client on the source. To build this
+client the developers need to provide
+
+1. A cloudevents source options (`options.CloudEventsSourceOptions`), this options have two parts
+    -  `sourceID`, it is a unique identifier for a source, for example, it can generate a source ID by hashing the hub
+    cluster URL and appending the controller name. Similarly, a RESTful service can select a unique name or generate a
+    unique ID in the associated database for its source identification.
+    - `CloudEventsOptions`, it provides cloudevents clients to send/receive cloudevents based on different event
+    protocol. We have supported the MQTT protocol (`mqtt.NewSourceOptions`), developers can use it directly.
+
+2. A resource lister (`generic.Lister`), it is used to list the resource objects on the source when resyncing the
+resources between sources and agents, for example, a hub controller can list the resources from the resource informers,
+and a RESTful service can list its resources from a database.
+
+3. A resource status hash getter method (`generic.StatusHashGetter`), this method will be used to calculate the resource
+status hash when resyncing the resource status between sources and agents.
+
+4. Codecs (`generic.Codec`), they are used to encode a resource object into a cloudevent and decode a cloudevent into a
+resource object with a given cloudevent data type. We have provided two data type (`io.open-cluster-management.works.v1alpha1.manifests`
+and `io.open-cluster-management.works.v1alpha1.manifestbundles`) for `ManifestWork`, they can be found in the `work/payload`
+package.
```

Comment:
nit: it would be better to add more description about how the event data type mapping to the cloudevent payload format.
For example:

For the `io.open-cluster-management.works.v1alpha1.manifests` data type, the cloud event payload contains a single resource object, while for `io.open-cluster-management.works.v1alpha1.manifestbundles`, it contains a list of resource objects.

Reply:
done